legend and widget examples

makella opened the pull request:
@VictorVelarde in this branch:

- I added a `legends` and `widgets` folder to examples
  - Legend examples: 
    - added ones from the guide + three from the `/misc` folder (category-chart, numeric-chart and linear-polygon)
    - did a few style/dataset modifications on the ones from the `/misc` folder
  - Widget examples: 
    - added the 3 examples from the Guide
- Formatted the `<head>` section of each one to be consistent
  - On the consistency note, I did not go through and comment in all the "new" steps for each example,
- Added the new folders + ordered examples in `examples.json`

To do:
- I still need clarity on how to do transparent legends... I haven't been able to get them to work yet. Asked in the issue for some help there.
- I think we will want to make sure this all works ok on the Dev Center, so I will go through the steps from before to see if I can get staging on this branch up and running

VictorVelarde commented:
@makella The ongoing work looks good to me.

I just found a little error: `classified-polygons` looks like a copy & paste duplication of 'category-points'.

I also fixed the broken urls to 'style.css' (one level less) with some other little adjustments (feel free to change / adapt):
- several examples: add zoom control
- 'category-line': replace 'CARTO_VL_OTHERS' with 'Others' in the legend
- 'linear-points': reduce decimals & reverse legend entries, to have north at the top.

makella commented:
ok, thank you @VictorVelarde! 

> I just found a little error: classified-polygons looks like a copy & paste duplication of category-points. 

OOPS

> I also fixed the broken urls to style.css (one level less) 

thank you! i was looking through the different examples and guides and couldn't find the exact standard since some varied... but now that you did this, i know that this is the pattern to follow so thank you!

> with some other little adjustments (feel free to change / adapt):
several examples: add zoom control
'category-line': replace 'CARTO_VL_OTHERS' with 'Others' in the legend
'linear-points': reduce decimals & reverse legend entries, to have north at the top.

all good ones! thank you for going through the details like this... super appreciate. :)

let me fix point one, test out the opacity on the legend, and get that example in and then i'll report back.

makella commented:
@VictorVelarde this is incomplete still, but added in the wording to the guide for transparent legends. 

I will fine-tune the following tomorrow:

to do for guide:
- [ ] re-read/edit
- [ ] clean-up final step map
- [ ] then add "step-x.html" for new transparent map features/legend example and link 

to do for examples:
- [ ] add transparent features/legend example
- [ ] I would like to replace the "other examples" in the guide with the modified versions you did... for example, the lat/long points with rounded numbers and north up

VictorVelarde commented on docs/guides/10-add-legends.md:
````diff
@@ -236,7 +222,70 @@ With that change, the final map will now label other categories as "Other weathe
+```
+At this point, if you refresh your map, you will see that the features on the map have the set opacity, but the features in the legend do not.
+
+If you want to add opacity to the legend items, you need to remove the `function rgbToHex(color)`, add constants for opacity (`const alpha`) and color (`const rgba`) where you can reference `alpha` for the `a` value of `const rgba`. Finally, you need to set the legend list items to draw from `const rgba`: 
````
IMO having so many details talking about `consts` and `alpha` in this paragraph looks a bit confusing, more than the code itself. 

Maybe we could say something like: 
*"If you want to add opacity to the legend items, you need to remove the `function rgbToHex(color)` (working just with the RGB color components of each entry) and use another approach, which includes the transparency. In this case, we have included that as a viz variable, named `myOpacity`, so it is just a matter of reading it when building our visual legend entries (thus building RGBA colors)."*

makella replied:
yep, that's the re-read/edit part! haha... i just kind of put it all down... your suggestions are great. will replace.

makella replied:
(you can also tell i don't know what i'm talking about LOL)

but yes, please feel free to comment on any part and i'll incorporate tomorrow.

i also removed one of the steps (step-2) as it was really uneeded, redundant. but all those changes will be clearer tomorrow

makella commented:
@VictorVelarde ok, I think I got everything... which is:

- [ ] updated Add Legends guide
  - added the opacity with variables section
  - added a new transparent legend example
  - replaced "Unclassed points" and "category line" examples with your updated versions

- [ ] added legends folder to examples section
  - added the examples from the guide plus a couple of other ones from the /misc/legends folder
  - updated examples.json with new section and examples
  
- [ ] added widgets folder to examples section
  - added examples from the guide
  - updated examples.json with new section and examples

I haven't done a check with the branch and Dev Center yet...

VictorVelarde commented:
Great! it would be nice to check them at staging. Ping me if you need any help for the deployment.

makella commented:
Yes! Will get that going today

VictorVelarde commented:
I have merged master and deployed this branch to our CDN, so the staging should be fixed now.

makella commented:
@VictorVelarde thank you! It is working

Been taking a look through, there are some things that I need to fix up nothing major, which is good

makella commented:
@VictorVelarde realized that the maps in the Widget guide and the examples that I took from there to put in the examples section all use the version below for Airship... I can change them manually now, but wondering if it is better to have updatable sources to the latest version like we do with VL.

```
    <!-- Airship -->
    <link rel="stylesheet" href="https://libs.cartocdn.com/airship-style/v1.0.0-beta.0/airship.css">
    <script src="https://libs.cartocdn.com/airship-components/v1.0.0-beta.0/airship.js"></script>
```

makella commented:
So far in this PR:
- [x] New examples for Legends and Widgets in Examples section
https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/#example-point-legend
- [x] For the examples that didn't come directly from Guides, I modified cartography and data source for a few of the legend/widget examples that were in the `misc` examples folder
- [x] Modified the height of histogram and chart for Examples section so they were better formatted for that view
https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/#example-category-chart
- [x] Updated all maps (in both guides and examples) where Airship beta was being referenced
https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/#example-viewport-histogram-widget
- [x] Updated Add widgets guide with proper Airship version
- [x] Added transparent legend workflow to Add legends guide
https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/guides/add-legends/#assign-opacity-with-variables

makella commented:
@VictorVelarde not sure why all my checks weren't successful... has to do with my last commit where I updated the Airship reference in the Add widgets guide. Any ideas?

VictorVelarde commented:
Hi @makella. The error (1 failed e2e test: timerange) looks unrelated to that update. Editing the .md should not be affecting us that way. We'll have to check that one in more detail...

makella commented:
Also added:

- [ ] basemap sections to examples with the following:
  - [labels under](https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/#example-basemap-labels-under)
  - [labels on top](https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/#example-basemap-labels-on-top)
  - [change background color](https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/#example-color-map-background) (basically no basemap)
  - [project layer](https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/#example-project-layer)
  - [set pitch and bearing](https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/#example-set-pitch-and-bearing)

@rochoa for the more advanced basemap layer ordering, I want to take a little time to better understand what jaak + team did and why (I need to do this anyway for vector basemaps so it will be a good exercise)

With these examples + legends + widgets, you can see that the examples section is getting more robust! There is still many to-dos but nice to see the expanding list :)
https://vl-dev-updates.developers.carto-staging.com/developers/carto-vl/examples/

@VictorVelarde if everything looks ok, it would be great to merge! 

thanks!

VictorVelarde commented:
I have fixed some issues and merged the PR. @makella have a look at these opacity changes [here](https://github.com/CartoDB/carto-vl/pull/1184/commits/bd31e0ae91813c8dcc16de09ebbc4672e9506308)

makella commented:
thanks!!
